src/conversions/array.rs

```diff
@@ -0,0 +1,276 @@
+    26, 27, 28, 29, 30, 31, 32
+);
+
+#[cfg(all(min_const_generics, not(feature = "nightly")))]
+impl<'a, T, const N: usize> FromPyObject<'a> for [T; N]
+where
+    T: FromPyObject<'a>,
+{
+    #[cfg(not(feature = "nightly"))]
+    fn extract(obj: &'a PyAny) -> PyResult<Self> {
+        create_array_from_obj(obj)
+    }
+
+    #[cfg(feature = "nightly")]
+    default fn extract(obj: &'a PyAny) -> PyResult<Self> {
+        create_array_from_obj(obj)
+    }
+}
+
+#[cfg(not(min_const_generics))]
+macro_rules! array_impls {
+    ($($N:expr),+) => {
+        $(
+            impl<T> IntoPy<PyObject> for [T; $N]
+            where
+                T: ToPyObject
+            {
+                fn into_py(self, py: Python) -> PyObject {
+                    self.as_ref().to_object(py)
+                }
+            }
+        )+
+    }
+}
+
+#[cfg(not(min_const_generics))]
+array_impls!(
+    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25,
+    26, 27, 28, 29, 30, 31, 32
+);
+
+#[cfg(min_const_generics)]
+impl<T, const N: usize> IntoPy<PyObject> for [T; N]
+where
+    T: ToPyObject,
+{
+    fn into_py(self, py: Python) -> PyObject {
+        self.as_ref().to_object(py)
+    }
+}
+
+#[cfg(all(min_const_generics, feature = "nightly"))]
+impl<'source, T, const N: usize> FromPyObject<'source> for [T; N]
+where
+    for<'a> T: FromPyObject<'a> + crate::buffer::Element,
+{
+    fn extract(obj: &'source PyAny) -> PyResult<Self> {
+        let mut array: core::mem::MaybeUninit<[T; N]> = core::mem::MaybeUninit::uninit();
+        // first try buffer protocol
+        if unsafe { crate::ffi::PyObject_CheckBuffer(obj.as_ptr()) } == 1 {
+            if let Ok(buf) = crate::buffer::PyBuffer::get(obj) {
+                if buf.dimensions() == 1 && buf.copy_to_slice(obj.py(), &mut array).is_ok() {
+                    buf.release(obj.py());
+                    // SAFETY: The array should be fully filled by `copy_to_slice`
+                    return Ok(unsafe { array.assume_init() });
+                }
+                buf.release(obj.py());
+            }
+        }
+        // fall back to sequence protocol
+        extract_sequence_into_slice(obj, &mut array)?;
+        // SAFETY: The array should be fully filled by `extract_sequence_into_slice`
+        Ok(unsafe { array.assume_init() })
+    }
+}
+
+// Helper to safely create arrays since the standard library doesn't
+// provide one yet. Shouldn't be necessary in the future.
+#[cfg(min_const_generics)]
+struct ArrayGuard<'a, T, const N: usize> {
+    dst: *mut T,
+    initialized: &'a mut usize,
+}
+
+#[cfg(min_const_generics)]
+impl<T, const N: usize> Drop for ArrayGuard<'_, T, N> {
+    fn drop(&mut self) {
+        debug_assert!(*self.initialized <= N);
+        let initialized_part = core::ptr::slice_from_raw_parts_mut(self.dst, *self.initialized);
+        unsafe {
+            core::ptr::drop_in_place(initialized_part);
+        }
```

Comment:
Why do we need to `drop` the initialized slice?

Reply:
Guard drops the initialized part when the closure panics to prevent leaks

Reply:
This data structure was used by previous releases of the nightly compiler and is battle tested by many people around the globe. Please consider avoid changing its original formulation.

Reply:
@kngwyu note that eventually an API for this might be added to `std`: https://github.com/rust-lang/rust/pull/75644

Personally I'm happy as long as we're sure the implementation we use is sound.